Staged ledger hash optional for dump-ledger; script to generate OCaml genesis

psteckler opened the pull request:
The CLI command `advanced dump-ledger` required a stage ledger hash as an anonymous argument. Since `client status` no longer provides that, hard to know what that hash is. With this PR, the hash is optional, given by the flag `-staged-ledger-hash`. (Currently, it must be in sexp format, which I'm guessing is decimal; see issue #3036.) The command now uses the same mechanism to obtain a hash that `status` used to use to supply a default if the hash is not given.

Also, there's Python script that runs `advanced dump-ledger` and prints out OCaml source for a genesis ledger. Output is to stdout, so you can choose to redirect that to whatever file you'd like.

Tested by running daemon locally, using time offset to obviate bootstrapping, running script to create new genesis file; then rebuilt coda using that genesis file.

Closes #2935.

nholland94 commented on scripts/genesis-from-ledger.py:
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python
```
I would recommend using a jinja template for this. It will make this much easier to read, and we already rely on jinja templates for generating CI configurations. http://jinja.pocoo.org/

psteckler replied:
I think that's a bit of overkill for this small piece of code.

I'll make it easier to read, by putting the pre/post/lude into arrays, and printing those in a loop.

bkase replied:
I agree with @nholland94 here, but willing to let this go in as is

bkase commented on src/lib/coda_lib/coda_lib.ml:
```diff
@@ -188,7 +188,17 @@ let best_protocol_state = compose_of_option best_protocol_state_opt
+            staged_ledger |> Staged_ledger.hash
+        | `Bootstrapping ->
+            failwith
+              "get_ledger: can't get staged ledger hash while bootstrapping" )
```
we should never be able to crash the daemon triggered via an RPC call -- can we put this back to an `Or_error`?

psteckler replied:
I'll try ...

psteckler commented:
Changed the code generator to use Jinja, added a Makefile target `genesis-ledger-ocaml`.

(You'll have to remove the `INFO Not using Docker` thing the Makefile generates when redirecting the output ...)

bkase commented on .genesis-ledger.ml.jinja:
```diff
@@ -0,0 +1,24 @@
+     {{- '; ' + account }}
+   {%- endif %}
+{%- endfor %}
+]
```
this is much cleaner!

psteckler replied:
my hair is grayer, though
